File: backend/src/services/directory.service.js
const labModel = require('../models/lab.model');
const testModel = require('../models/test.model');

async function withTests(labs) {
  return Promise.all(
    labs.map(async (lab) => ({ ...lab, tests: await testModel.findByLabId(lab._id) })),
  );
}

async function getFeaturedLabs() {
  const labs = await labModel.findFeatured();
  return withTests(labs);
}

// [type] is the partner's own declared category (from the "Services
// Offered" list at registration, e.g. "Diagnostics"/"Pharmacy") - matched
// against the real `services` a lab registered with, not a guess.
async function getApprovedLabs({ type, q, state } = {}) {
  const labs = q ? await labModel.search(q) : await labModel.findApproved();
  let results = await withTests(labs);
  if (type && type !== 'all') {
    results = results.filter((lab) => (lab.services || []).includes(type));
  }
  if (state && state !== 'all') {
    results = results.filter((lab) => lab.state === state);
  }
  return results;
}

function getLabTests(labId) {
  return testModel.findByLabId(labId);
}

function getTrendingTests() {
  return testModel.findTrending();
}

module.exports = { getFeaturedLabs, getApprovedLabs, getLabTests, getTrendingTests };

File: mobile/lib/services/directory_service.dart
import '../models/directory_models.dart';
import 'api_client.dart';

class DirectoryService {
  DirectoryService(this._api);

  final ApiClient _api;

  Future<List<Lab>> getFeaturedLabs() => _api.get(
    '/directory/labs',
    (data) => (data['labs'] as List<dynamic>)
        .map((l) => Lab.fromJson(l as Map<String, dynamic>))
        .toList(),
    query: {'featured': 'true'},
  );

  /// All admin-approved labs, not just the curated "featured" subset -
  /// this is what makes an approved partner actually discoverable.
  /// [type] filters by a partner's declared category (e.g. "Diagnostics"),
  /// [query] performs a real backend name search, [state] filters by the
  /// partner's registered state - all forwarded straight to the API.
  Future<List<Lab>> getApprovedLabs({
    String? type,
    String? query,
    String? state,
  }) => _api.get(
    '/directory/labs',
    (data) => (data['labs'] as List<dynamic>)
        .map((l) => Lab.fromJson(l as Map<String, dynamic>))
        .toList(),
    query: {
      if (type != null && type != 'all') 'type': type,
      if (query != null && query.trim().isNotEmpty) 'q': query.trim(),
      if (state != null && state != 'all') 'state': state,
    },
  );

  Future<List<LabTest>> getTrendingTests() => _api.get(
    '/directory/tests/trending',
    (data) => (data['tests'] as List<dynamic>)
        .map((t) => LabTest.fromJson(t as Map<String, dynamic>))
        .toList(),
  );
}
